protocols/gossipsub/src/behaviour.rs

```diff
@@ -2985,13 +2989,34 @@ where
 
         // Remove the connection from the list
         // If there are no connections left, inject_disconnected will remove the mapping entirely.
-        if let Some(connections) = self.peer_protocols.get_mut(peer_id) {
-            if let Some(index) = connections
-                .connections
-                .iter()
-                .position(|v| v == connection_id)
-            {
-                connections.connections.remove(index);
+        let connections = self
+            .connected_peers
+            .get_mut(peer_id)
+            .expect("Peers connection must be registered");
```

Comment:
As far as I can tell, blacklisted peers are not added to `self.connected_peers`:

```rust

    fn inject_connected(&mut self, peer_id: &PeerId) {
        // Ignore connections from blacklisted peers.
        if self.blacklisted_peers.contains(peer_id) {
            debug!("Ignoring connection from blacklisted peer: {}", peer_id);
            return;
        }


    fn inject_connection_established(
        &mut self,
        peer_id: &PeerId,
        connection_id: &ConnectionId,
        endpoint: &ConnectedPoint,
    ) {
        // Ignore connections from blacklisted peers.
        if self.blacklisted_peers.contains(peer_id) {
            return;
        }
```

Thus panicking here is not safe, right?

Sorry in case [my comment](https://github.com/libp2p/rust-libp2p/pull/2043/#discussion_r625681749) was misleading. I was referring to the inner `if` only.

Reply:
Yep, good catch. 

I usually dont miss things like this, but am across a few things at the moment. :sweat_smile: